Sync up with the Plone roadmap

dataflake opened the pull request:
This PR adds more Plone release-specific information and synchronizes some Plone roadmap details with Zope. Please speak up and correct me if you see anything you don't agree with.

dataflake committed:
- sync up with the Plone roadmap

mauritsvanrees reviewed (approved):
This seems good to me. It fits what we have on the Plone release schedule currently.
We might edit the Plone schedule during the Plone conference this week.

icemac commented on docs/developer/roadmap.rst:
```diff
@@ -37,27 +49,20 @@ Zope 4 - previous version
     - Full support: ended on 2020-10-08 (the Zope 5.0 release date)
-    - Bug fixes: until 2021-12-31
-    - Security fixes: until 2022-12-31 [2]_
+    - Bug fixes: until 2021-12-31 or until Plone 6.1 is released, whichever
```
Uh, that's surprising: Plone 6 is not yet released but we now want to support Zope 4 until the release of its successor?! According to https://plone.org/download/release-schedule there are not even plans for 6.1 yet.

I was thinking we'll end the bug fix mode for Zope 4 when Plone 5.2 (the only version using it) drops into security fix only mode at the end of 2021. – Okay, I did not read the whole story on the roadmap, it says exactly that Plone 6.1 has to be releases beforehand.

That's a bit sad: We still have to support Python 2.7, 3.5 and 3.6 for all Zope dependencies. Each year a new Python version is released which has to be supported on top of the existing ones. – Yes, there are mostly no code changes needed but running the tests takes longer with each supported version. Newer Python versions will make it less easy to support such a wide range of versions.

This promise pushes the ability to reduce maintenance burden of unsupported Python versions far into the future. I cannot say what I like this.

dataflake replied:
I totally understand what you mean. My intent was to sync up claims on the Plone roadmap with our own in order to start a discussion. I even left out the fact that Plone releases <5.2 are apparently still supported as well, which means Zope 2.13 is supposedly supported, which is clearly untrue.

I do not mind drawing a line into the sand and setting our own fixed support end date for Zope 4. The Plone release map is clearly too vague to use it as a basis for our own plans.

Make a suggestion for a date and/or Plone version that signals the end of support. How about Bugfix support ends with the release of Plone 6.0 or the end of this year, whichever comes later? That probably pushes it to early spring of 2022. Security fixes could stay at the end of 2022.

mauritsvanrees replied:
I am the one who pushed for adding dates in the Plone release schedule early last year. It seemed best to make it a combination: support until year Y or until Plone X is released. And then the question was how to combine these two: whichever comes first or whichever comes last. For users of Plone it seemed fairer to say: whichever comes last.

Problem: Plone 6 only has a proper alpha release since one week. It has taken much longer than anticipated, mostly due to Covid and not being able to see each other in person. So we may need to rethink and rephrase our promises.

When the Plone schedule says a Plone version is supported, this does not necessarily mean that the Zope part is supported as well. If bugfix support for Zope 4 ends, I am happy to still support Plone 5.2 for longer than that, and make releases. Most bugs that we find by now, will be in core Plone, not in Zope. If really needed, we can do a patch in Plone.
Security support will be longer I hope, but most security fixes will likely be similar in Zope 4 and 5.
Again, if needed, Plone could do a patch of Zope in its own code, although a pair of Zope eyes on that would be nice.

Note that Plone 5.2 and Zope 4 are special, in that they are a required stepping stone if you want to migrate a ZODB from Python 2 to 3. That means it may make sense for both Plone and Zope to offer longer support, to give users time to make the leap. But that is more true for security support.
(Since this year there is `collective.exportimport`, which can export Plone 4.3 Py 2 and import it directly in Plone 6 Py 3. So the intermediate stepping stone may be less needed.)

I see logic in ending Zope 4 bugfix support when Plone 6.0 is released, which I now expect in February/March 2022. I may want to say for Plone 5.2 itself: bugfix/maintenance support for half a year longer. I expect that after half a year, work on Plone 6.1 will have started in earnest, and we cannot sustain three releases.

Ending Zope 4 security support at the end of 2022 sounds fair. A bit less than a year after release of Plone 6. For Plone 5.2 itself I would like it longer. If you have one year of security support overlap, that gives you only one year to migrate, which is not much when you need to migrate several clients. But February 2023 would already mean 7.5 years of security support since Plone 5.0, and 3.5 years since Plone 5.2. That is very long.
We said December 2023 or until Plone 7 comes out. I think we should forget about Plone 7. After Plone 6.0 comes out, 2 years security support would be what I hope we can offer on behalf of Plone.

What helps, is that in my opinion migrating from Plone 5.2 to 6.0 is relatively easy, as long as you stick to Classic Plone and not the new Volto frontend.

I am thinking out loud here. We will have a discussion during the conference, planned this Friday.

icemac replied:
Currently each commit made on Zope `master` (aka Zope 5) is ported back to Zope 4. In the dependent packages we keep supporting the Python versions which are needed for Zope 4 so we can easily update Zope 4 to use a new version of a dependency  to get a bugfix or even a new feature.

My idea would be to end this general approach - maybe when Plone 6 final is released – and switch to a more fine graded approach: If a commit made on `master` is actually needed on to fix Zope 4 behavior it gets backported. When the fix is provided it should be indicated that Plone 5.2 needs it, too (or another customer/consumer still running on Zope 4).

Currently we claim that an application which runs without deprecation warnings on Zope 4 using Python 3 can run unchanged on Zope 5. Using my suggestion from above this claim could break. (I do no have any current experience if we currently still fulfill it. The last time I switched a project from Zope 4 to Zope 5 was nearly exactly a year ago (4.5 -> 5.0) and it went without any problems.)

I personally have no problem with extending the security support as it is low volume and helps customers to stay save.

dataflake committed:
- adjust support policy to find a compromise

dataflake commented:
I just made the following changes:

- Zope 4 bugfix support ends at the end of 2021 or the release of Plone 6.0, whichever comes later
- Zope 4 security issues are fixed until the end of 2022, I removed any mentions of corresponding Plone releases
- added a note to Zope 2.13 to make it clear that it is unsupported, but in a Plone emergency new releases could appear

> Currently we claim that an application which runs without deprecation warnings on Zope 4 using Python 3 can run unchanged on Zope 5. Using my suggestion from above this claim could break. (I do no have any current experience if we currently still fulfill it. The last time I switched a project from Zope 4 to Zope 5 was nearly exactly a year ago (4.5 -> 5.0) and it went without any problems.)

My opinion: The claim should stand and if there are problems migrating from the latest Zope 4 to Zope 5 I would consider that a bug. I haven't tested such a migration in recent times, but don't want to break that promise. I don't recall any changes in Zope 5 that would lead to problems. If we plan on introducing changes that would cause a breakage then we should create a 5.x branch and consider `master` Zope 6, which would allow some breakage again.

mauritsvanrees commented on docs/developer/roadmap.rst:
```diff
@@ -37,27 +49,21 @@ Zope 4 - previous version
 Zope 2.13 - old version
 -----------------------
+Plone 4.3, 5.0 and 5.1 are running on top of Zope 2.13. Zope 2.13 is no longer
+officially supported but may see releases in response to secyrity issues
```
Typo: `secyrity` -> `security`.

mauritsvanrees commented:
> I just made the following changes:
> 
> * Zope 4 bugfix support ends at the end of 2021 or the release of Plone 6.0, whichever comes later
> * Zope 4 security issues are fixed until the end of 2022, I removed any mentions of corresponding Plone releases

Agreed.

> * added a note to Zope 2.13 to make it clear that it is unsupported, but in a Plone emergency new releases could appear

I would not expect any new releases, so no Zope 2.13.31. I don't do any releases for Plone 5.1 and lower either.
But a review of a hotfix package for Plone would be welcome in such a case.
(I sometimes think of releasing new versions for old Plone releases, where the only change is that they have a new dependency on a hotfix package, but that is not what we have done so far.)

> > Currently we claim that an application which runs without deprecation warnings on Zope 4 using Python 3 can run unchanged on Zope 5.
>
> My opinion: The claim should stand and if there are problems migrating from the latest Zope 4 to Zope 5 I would consider that a bug. I haven't tested such a migration in recent times, but don't want to break that promise. I don't recall any changes in Zope 5 that would lead to problems. If we plan on introducing changes that would cause a breakage then we should create a 5.x branch and consider `master` Zope 6, which would allow some breakage again.

Note that Plone does not have such a claim. When an add-on goes from 5.2 to 6.0, I expect its code will need some changes. Actually, this time the needed changes seem very small or non-existent so far, except for frontend code.
We try to keep the needed changes minimal for minor releases, although of course this utterly failed for 5.2 where we switched to Zope 4. But for a major release, things are allowed to break.

dataflake committed:
- typo and removal of release claims for Zope 2.13

dataflake commented:
> I would not expect any new releases, so no Zope 2.13.31. I don't do any releases for Plone 5.1 and lower either. But a review of a hotfix package for Plone would be welcome in such a case.

No problem, just assign me as reviewer. I removed the release claim.

About our claim that applications on Zope 4/Python 3 should port without issues to Zope 5: This claim is for Zope only. On the Zope side we have no control over backwards incompatibilities you're putting into Plone code.

mauritsvanrees commented:
I have a related question.
Say Plone 6.0.0 comes out early next year, and ships with Zope 5.3 which is at that point the latest Zope release.
A month later Zope 5.4 comes out.
Do you expect Plone 6.0.1 to upgrade to Zope 5.4 then?

For Plone it would be good if it could rely on bug fix releases of Zope 5.3.x for longer, say for a year, or until Plone 6.1 comes out (which I hope is roughly the same time).
But for the Zope team that would mean supporting three branches: 4.x, 5.3.x and 5.x.

During the Plone 5.2 series, we have had some updates of Zope that were no-brainers, no problem at all to upgrade. If that is the case for Zope 5.x, then it is no problem.
But some Zope 4 releases caused problems, especially the template-engine changes starting in Zope 4.4. In my opinion, they should have been postponed to Zope 5, which I can only say in hindsight, because the changes made sense at the time.

I am saying: if something like that happens in Zope 5.4, then Plone 6.0.x would want to stay on Zope 5.3.x.
It is of course also the responsibility of Plone to test with Zope checkout before it is actually released. Then we can give feedback.

dataflake commented:
We use semantic versioning. If a Zope 5.4 breaks software running on Zope 5.3 then that is a bug. (And no, I do *not* want to create branches for minor versions, period).

icemac commented:
In my option it would be helpful if Plone tries to stay on the latest Zope release even in bugfix versions (Plone 6.0.1 on Zope 5.4), especially if the update is a no-brainer.
This would allow us to support only major versions of Zope. I agree with Jens that supporting multiple minor versions with bugfixes is too much of a hassle.

@mauritsvanrees How do you handle update of other dependencies in bugfix releases? Do you update them to newer versions?

dataflake commented:
> In my option it would be helpful if Plone tries to stay on the latest Zope release even in bugfix versions (Plone 6.0.1 on Zope 5.4), especially if the update is a no-brainer. This would allow us to support only major versions of Zope.

I agree (my first answer left out a lot of details).

We had moments like you described when there were backwards incompatibilities within a release series, but that's still bad and we should not do that again. I believe Zope 4 and 5 have now reached a level of stability that will make following the convention of not breaking applications from one Zope release to the next easier.

You might have noticed this whenever we make a release of Zope 4 or 5: Immediately after the release we update all dependencies that can safely be updated. For Zope 5 that means all of them, Zope 4 is hamstrung by still having to support Python 2.7, which many dependencies have moved away from. That way those new dependencies get some testing before they are rolled out with the following Zope release. So with our releases we do update all dependencies for every release.

dataflake commented:
P.S.: I see two interests here that can be satisfied with a little discipline and coordination:

- The Plone release managers don't want surprises when updating the Zope dependency within the same release series.
- The Zope release managers want to limit support to major versions and not deal with bugfixes for minor releases

IMHO if the Plone side does testing during development or before a release with Zope pinned to the tip of the respective release branch then incompatibilities should become visible quickly. And we'd then promise to fix those issues in time for the Plone release.

mauritsvanrees commented:
> How do you handle update of other dependencies in bugfix releases? Do you update them to newer versions?

Yes, but irregularly. Usually I or @jensens will use `plone.versioncheck` (or `z3c.checkversions`) to check for newer versions, and then manually decide which ones seem safe to update. In Plone 6 it can usually be the latest version, on Plone 5.2 I am more conservative.
I think in Zope 4 you pin the latest version that is compatible with Python 2, right? In Plone in a few cases we use `[versions:python2]` and `[versions:python3]` to have a newer version on Python 3.

I think at some point both Zope and Plone should move to using a `constraints.txt` as the canonical place for version pins. I assume there are tools in pip-land that are better at figuring out which versions to use for which Python version, which may become important when packages start dropping support for older Python version with something like `requires_python=>=3.8`, which buildout happily ignores, or at least buildout 2. But I have not found a proper workflow for that yet. See also [community.plone.org](https://community.plone.org/t/plone-6-0-0a1-released/14431/11?u=mauritsvanrees).

> IMHO if the Plone side does testing during development or before a release with Zope pinned to the tip of the respective release branch then incompatibilities should become visible quickly. And we'd then promise to fix those issues in time for the Plone release.

Yes, that would be good, and we should have done that sooner. I will make a PR on our side.

dataflake commented:
> I think at some point both Zope and Plone should move to using a constraints.txt as the canonical place for version pins.

We already use that. Our workflow for updating dependency pins is as follows:

- we use `z3c.checkversions` to suggest version updates and add them to `versions.cfg` and `versions-prod.cfg`
- we run `bin/buildout` which will automatically transcribe those version numbers to `requirements.txt` and `constraints.txt`

At the end all those files are in sync.

mauritsvanrees commented:
What I mean with constraints as canonical location, is that we update `constraints.txt` and then `versions.cfg` is derived from that, not the other way around.
But so far, the current process works, and we have something similar in Plone.

dataflake commented:
I'll merge and publish this now. I'm glad how this has started a discussion about backwards compatibility and how the two projects can coordinate better. It provided me with some good new guidelines.
